File: polybar/.config/polybar/modules/weather.php
<?php
define("CITY", "Miami");
define("LATITUDE", 25.75);
define("LONGITUDE", -80.25);
define("TEMPERATURE_UNIT", "celsius");
define("TIMEZONE", "EST");
$url = join("?", [
    "https://api.open-meteo.com/v1/forecast",
    http_build_query([
        "current_weather" => "true",
        "latitude" => LATITUDE,
        "longitude" => LONGITUDE,
        "temperature_unit" => TEMPERATURE_UNIT,
        "timezone" => TIMEZONE
    ], "", "&")
]);
$contents = file_get_contents($url);
if (empty($contents)) {
    printf("󰅤");
    return;
}
$json = json_decode($contents);
if (property_exists($json, "error") && $json->error) {
    printf($json->reason);
    return;
}
$codes = array(
    0 => array(
        "label" => "Clear",
        "day" => "󰖙",
        "night" => "󰖔"
    ),
    1 => array(
        "label" => "Mainly Clear",
        "day" => "󰖙",
        "night" => "󰖔"
    ),
    2 => array(
        "label" => "Partly Cloudy",
        "day" => "󰖕",
        "night" => "󰼱"
    ), 
    3 => array(
        "label" => "Overcast",
        "day" => "󰖐",
        "night" => "󰖐"
    ),
    45 => array(
        "label" => "Fog",
        "day" => "󰖑",
        "night" => "󰖑"
    ),
    48 => array(
        "label" => "Depositing Rime Fog",
        "day" => "󰖑",
        "night" => "󰖑"
    ),
    51 => array(
        "label" => "Light Drizzle",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    53 => array(
        "label" => "Moderate Drizzle",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    55 => array(
        "label" => "Dense Drizzle",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    56 => array(
        "label" => "Light Freezing Drizzle",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    57 => array(
        "label" => "Dense Freezing Drizzle",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    61 => array(
        "label" => "Slight Rain",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    63 => array(
        "label" => "Moderate Rain",
        "day" => "󰖖",
        "night" => "󰖖"
    ),
    65 => array(
        "label" => "Heavy Rain",
        "day" => "󰖖",
        "night" => "󰖖"
    ),
    66 => array(
        "label" => "Light Freezing Rain",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    67 => array(
        "label" => "Heavy Freezing Rain",
        "day" => "󰖖",
        "night" => "󰖖"
    ),
    71 => array(
        "label" => "Slight Snow",
        "day" => "󰖘",
        "night" => "󰖘"
    ),
    73 => array(
        "label" => "Moderate Snow",
        "day" => "󰼶",
        "night" => "󰼶"
    ),
    75 => array(
        "label" => "Heavy Snow",
        "day" => "󰼶",
        "night" => "󰼶"
    ),
    77 => array(
        "label" => "Snow Grains",
        "day" => "󰖘",
        "night" => "󰖘"
    ),
    80 => array(
        "label" => "Slight Rain Showers",
        "day" => "󰖗",
        "night" => "󰖗"
    ),
    81 => array(
        "label" => "Moderate Rain Showers",
        "day" => "󰖖",
        "night" => "󰖖"
    ),
    82 => array(
        "label" => "Heavy Rain Showers",
        "day" => "󰖖",
        "night" => "󰖖"
    ),
    85 => array(
        "label" => "Slight Snow Showers",
        "day" => "󰖘",
        "night" => "󰖘"
    ),
    86 => array(
        "label" => "Heavy Snow Showers",
        "day" => "󰼶",
        "night" => "󰼶"
    )
);
$temperature_units = array(
    "celsius" => "C",
    "fahrenheit" => "F"
);
list("current_weather" => $current_weather) = get_object_vars($json);
$code = (object) $codes[$current_weather->weathercode] ?? null;
$components = array(
    "city" => CITY,
    "weather_icon" => date("H") <= 18 ? $code?->day : $code?->night,
    "weather_label" => $code?->label,
    "temp" => $current_weather->temperature,
    "temp_unit" => $temperature_units[TEMPERATURE_UNIT]
);
printf(
    "%s %s %s°%s",
    $components["weather_icon"],
    $components["weather_label"],
    $components["temp"],
    $components["temp_unit"]
);

File: awesome/.config/awesome/user/widgets/weather.lua
local awful = require("awful")
local icons = require("user.theme.icons")
local wibox = require("wibox")

local Weather = {}
Weather.__index = Weather

Weather.xref = {
    [0] = { "Clear Sky", icons.weather_day_sunny, icons.weather_night_clear },
    [1] = { "Mainly Clear", icons.weather_day_sunny, icons.weather_night_clear },
    [2] = { "Partly Cloudy", icons.weather_day_cloudy, icons.weather_night_cloudy },
    [3] = { "Overcast", icons.weather_day_cloudy, icons.weather_night_cloudy },
    [45] = { "Fog", icons.weather_day_fog, icons.weather_night_fog },
    [48] = { "Fog", icons.weather_day_fog, icons.weather_night_fog },
    [51] = { "Drizzle", icons.weather_day_rain, icons.weather_night_rain },
    [53] = { "Drizzle", icons.weather_day_rain, icons.weather_night_rain },
    [55] = { "Drizzle", icons.weather_day_rain, icons.weather_night_rain },
    [56] = { "Freezing Drizzle", icons.weather_day_rain, icons.weather_night_rain },
    [57] = { "Freezing Drizzle", icons.weather_day_rain, icons.weather_night_rain },
    [61] = { "Rain", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [63] = { "Rain", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [65] = { "Rain", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [66] = { "Freezing Rain", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [67] = { "Freezing Rain", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [71] = { "Snow", icons.weather_day_snow, icons.weather_night_snow },
    [73] = { "Snow", icons.weather_day_snow, icons.weather_night_snow },
    [75] = { "Snow", icons.weather_day_snow, icons.weather_night_snow },
    [77] = { "Snow Grains", icons.weather_day_snow, icons.weather_night_snow },
    [80] = { "Rain Showers", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [81] = { "Rain Showers", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [82] = { "Rain Showers", icons.weather_day_thunderstorm, icons.weather_night_thunderstorm },
    [85] = { "Snow Showers", icons.weather_day_snow, icons.weather_night_snow },
    [86] = { "Snow Showers", icons.weather_day_snow, icons.weather_night_snow }
}

function Weather:new(args)
    local weather = setmetatable({}, Weather)
    weather.latitude = args and args.latitude or 52.52
    weather.longitude = args and args.longitude or 13.41
    local icon = wibox.widget.textbox()
    weather.__icon = icon
    icon.font = "FiraCode Nerd Font Mono 18"
    icon.text = icons.weather_day_sunny
    local spacing = wibox.container.margin()
    spacing.right = 8
    local desc = wibox.widget.textbox()
    weather.__desc = desc
    desc.text = "Clear Sky"
    local temp = wibox.widget.textbox()
    weather.__temp = temp
    temp.text = "0.0°"
    local layout = wibox.layout.fixed.horizontal()
    layout:add(icon)
    layout:add(spacing)
    layout:add(desc)
    layout:add(spacing)
    layout:add(temp)
    local container = wibox.container.margin(layout)
    weather.__widget = container
    container.left = 8
    container.right = 8
    timer({
        autostart = true,
        callback = function ()
            weather:update()
        end,
        timeout = args and args.timeout or 120
    })
    weather:update()
    return weather
end

function Weather:update()
    local url = string.format(
        "https://api.open-meteo.com/v1/forecast?latitude=%s&longitude=%s&current_weather=true",
        self.latitude,
        self.longitude
    )
    awful.spawn.easy_async(string.format("curl '%s'", url), function (output)
        self.temperature = output:match("\"temperature\":([%d%.]+)")
        self.code = tonumber(output:match("\"weathercode\":(%d+)"))
        self:refresh()
    end)
    collectgarbage("collect")
end

function Weather:refresh()
    local index = self.xref[self.code] and self.code or 1
    local desc, icon0, icon1 = table.unpack(self.xref[index])
    local hour = tonumber(os.date("%H"))
    local daylight = hour >= 6 and hour < 18
    self.__icon.text = daylight and icon0 or icon1
    self.__desc.text = desc
    self.__temp.text = string.format("%s°", self.temperature)
end

return setmetatable(Weather, {
    __call = function (_, ...)
        local weather = Weather:new(...)
        return weather.__widget
    end
})
